### PR TITLE
Integrate multi_tensor zero_grad into Optimizer base class

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* #69982
* #69981
* #69980
* __->__ #69936

Currently, the optimizers in `torch/optim/_multi_tensor/` all override the base Optimizer class' implementation of `zero_grad` with the same foreach zero_grad implementation (e.g. [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/adadelta.py#L93-L114)). There is a TODO that indicates that this should be refactored to the base class once the foreach ops are in good shape. This PR is intended to address that TODO.

Differential Revision: [D33346748](https://our.internmc.facebook.com/intern/diff/D33346748)

### CODE REVIEW

File: torch/optim/optimizer.py [hunk 1]
```diff
@@ -201,8 +201,12 @@ def zero_grad(self, set_to_none: bool = False):
                 (in one case it does the step with a gradient of 0 and in the other it skips
                 the step altogether).
         """
+        foreach = self.defaults['foreach'] if ('foreach' in self.defaults) else False
```
Comment (albanD) [inline torch/optim/optimizer.py:204]: `foreach = self.defaults.get('foreach', False)`

File: torch/optim/optimizer.py [hunk 2]
```diff
@@ -214,7 +218,14 @@ def zero_grad(self, set_to_none: bool = False):
                                 p.grad.detach_()
                             else:
                                 p.grad.requires_grad_(False)
-                            p.grad.zero_()
+                            if (not foreach or p.grad.is_sparse):
+                                p.grad.zero_()
+                            else:
+                                per_device_and_dtype_grads[p.grad.device][p.grad.dtype].append(p.grad)
+            if (foreach):
```
Comment (albanD) [inline torch/optim/optimizer.py:225]: nit no parenthesis

**Review by albanD — COMMENTED** [review]
change looks good. just some nits for the impl

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33346748).

**Review by albanD — APPROVED** [review]

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33346748).

**Pull request closed without merging**